python/src/iceberg/catalog/in_memory.py

```diff
@@ -0,0 +1,106 @@
+#  Licensed to the Apache Software Foundation (ASF) under one
+#  or more contributor license agreements.  See the NOTICE file
+#  distributed with this work for additional information
+#  regarding copyright ownership.  The ASF licenses this file
+#  to you under the Apache License, Version 2.0 (the
+#  "License"); you may not use this file except in compliance
+#  with the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing,
+#  software distributed under the License is distributed on an
+#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+#  KIND, either express or implied.  See the License for the
+#  specific language governing permissions and limitations
+#  under the License.
+
+from typing import Dict, List, Optional
+
+from iceberg.catalog.base import AlreadyExistsError, Catalog, TableNotFoundError
+from iceberg.table.base import Table, TableSpec
+
+
+class InMemoryCatalog(Catalog):
```

Comment:
Thanks for adding this! This will be great for demos/quickstarts/tests. If it's not too much would you mind breaking this out into a separate PR? I think there's value in getting just the catalog interface merged in asap so different implementations can be worked on in parallel.

Reply:
I am done with the In-memory catalog implementation and have committed it within this PR. I think it is important to test the Catalog interface with a concrete implementation to see if it holds firm, and will also help the reviewer visualize the interface in action better. I'll mark this PR ready for review now :)

Reply:
The in memory implementation should only exist in the test srcs so that it does not get confused / used for a real implementation.  Also echo Sam's comment that splitting the PR will make it easier to get the catalog in.  We can always follow up with changes if necessary.